Added UnsortedBasecallsConverter.

jacarey opened the pull request:
Added a simple single threaded unsorted basecalls converter.

Commit:
Added UnsortedBasecallsConverter.

tfenne commented on src/main/java/picard/illumina/BasecallsConverterBuilder.java:
```diff
@@ -71,6 +71,17 @@ public BasecallsConverterBuilder(final File basecallsDir, final Integer lane, fi
+     *
+     * @return A basecalls converter that will output unsorted records.
+     */
+    public BasecallsConverter<CLUSTER_OUTPUT_RECORD> build() {
```
How would you feel about adding a method something like:

````java
public ... withSorting(final Comparator<CLUSTER_OUTPUT_RECORD> outputRecordComparator,
                       final SortingCollection.Codec<CLUSTER_OUTPUT_RECORD> codecPrototype,
                       final Class<CLUSTER_OUTPUT_RECORD> outputRecordClass,
                       final List<File> tmpDirs)
```

and then going back to having a single `build()` method?  I think we could even add to the builder a `withMaxRecordsInRam(int max)`  (with some sensible default if it's not called) so that is no longer in the `build()` signature either.  

I think it would be ideal if the builder took in the maximum as just a total maximum and then was allowed to figure out how to use it.  I.e. for a sorting version it should figure that each thread will have a tile, so set max per tile to `max / threads` or something like that.  And then the non-sorting one can use that to configure how big the buffers should be on the output writers when we move to the async pool?  Or is that done somewhere else?

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToFastq.java:
```diff
@@ -176,6 +176,9 @@
             "is larger than the available memory.")
     public Boolean FORCE_GC = true;
 
+    @Argument(doc = "If true, the output records are sorted by read name. Otherwise they are unsorted.")
```
I think we should add, either here (and in IBTS) or in the general usage that setting `SORT=false` doesn't produce unsorted data, but that it produces data in the order that the seuqencer (and bcl2fastq2) produce it - i.e. ordered by tile and within tile by position, but not with a full queryname sort.

tfenne commented on src/main/java/picard/illumina/UnsortedBasecallsConverter.java:
```diff
@@ -0,0 +1,95 @@
+ * from standard Illumina formats to specific output records (FASTA records/SAM records). This data is processed
+ * on a tile by tile basis.
+ * <p>
+ * The underlying IlluminaDataProvider apply several optional transformations that can include EAMSS filtering,
```
```suggestion
 * The underlying IlluminaDataProvider applies several optional transformations that can include EAMSS filtering,
```

tfenne commented on src/main/java/picard/illumina/UnsortedBasecallsConverter.java:
```diff
@@ -0,0 +1,95 @@
+public class UnsortedBasecallsConverter<CLUSTER_OUTPUT_RECORD> extends BasecallsConverter<CLUSTER_OUTPUT_RECORD> {
+    private static final Log log = Log.getInstance(UnsortedBasecallsConverter.class);
+    private final ProgressLogger readProgressLogger = new ProgressLogger(log, 1000000, "Read");
+    private final ProgressLogger writeProgressLogger = new ProgressLogger(log, 1000000, "Write");
```
I'd argue we only need one progress logger - the fact that we may not write PF reads isn't that interesting.  Maybe just change to `Processed` and have one?

tfenne reviewed (approved):
Looks good to me with one major suggestion to try and keep to a single `build()` method that doesn't take any parameters.

tfenne committed:
Speed up encoding of Casava read names.

tfenne committed:
Fixed a couple of little bugs.

tfenne committed:
With a simple test.

tfenne committed:
Attempt to make things much more efficient when not sorting by bypassing the conversion from ClusterData into FastqRecordsForCluster.

tfenne committed:
More performance improvements after some benchmarking.

Commit:
Merge pull request #5 from fulcrumgenomics/tf_casava_naming_speedup

Speed up encoding of Casava read names.

tfenne committed:
Merge pull request #6 from fulcrumgenomics/tf_dont_convert_so_much

Attempt to make things much more efficient when not sorting by bypass…

tfenne committed:
BclReader now reads from it's streams into a small buffer to speed things up.

tfenne committed:
Used pre-generated lookups for decoding BCL bases and quals.

tfenne committed:
A little more optimization.

tfenne committed:
Have BCL readers produce `N` instead of `.` internally to save having to change it later.

Commit:
Merge pull request #7 from fulcrumgenomics/tf_more_performance_tweaks

Another ~33% increase in single-thread performance for IlluminaBasecallsToFastq

Commit:
Review changes.
Added ClusterDataCodec so that SortedBasecallsConverter can use ClusterWriter instead of FastqRecordWriter for SortingCollections.

Commit:
Added basic write threading to sorted converter.

Commit:
Added basic write threading to unsorted converter.

Commit:
Small threading tweaks.

tfenne committed:
Fix to ensure that IOUtil.openFileForWriting() will respect CommandLineProgram.COMPRESSION_LEVEL when writing gzipped files.

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToFastq.java:
```diff
@@ -159,11 +163,13 @@
+    @Deprecated
     @Argument(doc = "Configure SortingCollections to store this many records before spilling to disk. For an indexed" +
-            " run, each SortingCollection gets this value/number of indices.")
+            " run, each SortingCollection gets this value/number of indices. Deprecated: use `MAX_RECORDS_IN_RAM`")
```
Given the deprecation should we default this to -1 or something so we can tell which one to use?

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToFastq.java:
```diff
@@ -632,77 +508,138 @@ public void close() {
+            boolean pf = this.binaryCodec.readBoolean();
+            String matchedBarcode = this.binaryCodec.readLengthAndString(false);
+
+            ClusterData clusterData = new ClusterData(readData);
```
Why not just declare this at the top of the method and set things directly into it?  That first try/catch could also be simplified then to not have to declare lane outside the try, with:
```java
try { clusterData.setLane(this.binaryCodec.readInt()); }
catch (final RuntimeEOFException e) { return null; }
```

jacarey replied:
I can move it up a bit, but since `ReadData` is private final I need to at least parse `numReads` so I can instantiate the ReadData to set in the constructor.

Alternatively I could open up access to `ReadData` and not have it be final in `ClusterData`

jacarey replied:
```
            int numReads = this.binaryCodec.readInt();

            ReadData[] readData = new ReadData[numReads];
            ClusterData clusterData = new ClusterData(readData);
            clusterData.setLane(this.binaryCodec.readInt());
            clusterData.setTile(this.binaryCodec.readInt());
            clusterData.setX(this.binaryCodec.readInt());
            clusterData.setY(this.binaryCodec.readInt());
            clusterData.setPf(this.binaryCodec.readBoolean());
            String matchedBarcode = this.binaryCodec.readLengthAndString(false);
            if (matchedBarcode.length() == 0) {
                clusterData.setMatchedBarcode(null);
            } else {
                clusterData.setMatchedBarcode(matchedBarcode);
            }
```

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToFastq.java:
```diff
@@ -632,77 +508,138 @@ public void close() {
+                }
+            }
+
+            this.binaryCodec.writeBoolean(clusterData.isPf());
+            if (clusterData.getMatchedBarcode() != null) {
+                this.binaryCodec.writeString(clusterData.getMatchedBarcode(), true, true);
+            } else {
+                this.binaryCodec.writeString("", true, false);
+            }
```
Unless there's a good reason I think it would be nicer to write all the 1-per-cluster items first before the reads, and then the reads.

tfenne commented on src/main/java/picard/illumina/IlluminaBasecallsToFastq.java:
```diff
@@ -632,77 +508,138 @@ public void close() {
         @Override
-        public SortingCollection.Codec<FastqRecordsForCluster> clone() {
-            return new FastqRecordsForClusterCodec(numTemplates, numSampleBarcodes, numMolecularBarcodes);
+        public int compare(ClusterData o1, ClusterData o2) {
+            return readNameEncoder.generateReadName(o1, 1).compareTo(readNameEncoder.generateReadName(o2, 1));
```
Ugh, kinda sucks to have to generate the read name each time, but I see why.  Maybe we should at least add a method to ReadNameEncoder to generate the name without the stuff after the space since that'd be a bit more efficient?

tfenne replied:
The other option would be to have a variable in ClusterData to hold a read name, then set it and compare it?

jacarey replied:
Since ClusterData holds more than one read we'd either have to generate them all or a prefix and then add to it based on the pairNumber, In this case we use the first pair for comparison only. So we could do that as well.

That would also mean more data written for the sorting collection when it spills to disk.

tfenne reviewed (approved):
This is looking really nice.  I didn't look closely at the UnsortedBasecallsConverter changes since I figure those will change again once the AsyncWriterPool is ready and I can take another look then.

Commit:
Review changes.
